src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContextSerializer.cs

```diff
@@ -25,25 +25,165 @@ private DesigntimeLicenseContextSerializer()
+                    buffer[0] = (byte)firstSByte;
+                    return _stream.Read(buffer, offset + 1, count - 1) + 1;
+                }
+                return _stream.Read(buffer, offset, count);
+            }
+
+            public override long Seek(long offset, SeekOrigin origin) => _stream.Seek(offset, origin);
+
+            public override void SetLength(long value) => _stream.SetLength(value);
+
+            public override void Write(byte[] buffer, int offset, int count) => _stream.Write(buffer, offset, count);
+
+            public override int ReadByte()
+            {
+                sbyte read = (sbyte)_stream.ReadByte();
+                firstSByte = read;
+                return read;
+            }
+        }
+
+        /// <summary>
+        /// During deserialization, the stream passed in may be binary formatted or may have used binary writer. This is quick test to discern between them.
+        /// </summary>
+        internal static bool StreamIsBinaryFormatted(StreamWrapper stream)
+        {
+            // For binary formatter, the first byte is the SerializationHeaderRecord and has a value 0
+            int firstByte = stream.ReadByte();
+            if (firstByte != 0)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static void DeserializeUsingBinaryFormatter(StreamWrapper wrappedStream, string cryptoKey, RuntimeLicenseContext context)
         {
+            try
+            {
 #pragma warning disable SYSLIB0011 // Issue https://github.com/dotnet/runtime/issues/39293 tracks finding an alternative to BinaryFormatter
-            IFormatter formatter = new BinaryFormatter();
+                IFormatter formatter = new BinaryFormatter();
 
-            object obj = formatter.Deserialize(o);
+                object obj = formatter.Deserialize(wrappedStream);
 #pragma warning restore SYSLIB0011
 
-            if (obj is object[] value)
+                if (obj is object[] value)
+                {
+                    if (value[0] is string && (string)value[0] == cryptoKey)
+                    {
+                        context._savedLicenseKeys = (Hashtable)value[1];
+                    }
+                }
+            }
+            catch (NotSupportedException exception)
+            {
+                if (!LocalAppContextSwitches.BinaryFormatterSerializationEnabled)
+                {
+                    throw new NotSupportedException(exception.Message + " Turn on the EnableUnsafeBinaryFormatterSerialization flag to continue using BinaryFormatter");
+                }
+                else
+                {
+                    throw exception;
+                }
+            }
+        }
+
+        internal static void Deserialize(Stream o, string cryptoKey, RuntimeLicenseContext context)
+        {
+            StreamWrapper wrappedStream = new StreamWrapper(o);
+            if (StreamIsBinaryFormatted(wrappedStream))
```

Comment:
I am not using the `AllowBinaryFormatter...` feature switch here, so we won't stub this part out by default. I think this is better UX in this particular case because I expect all `stream`s passed in to this method currently are binary formatted.

Reply:
I think we should be checking the AppContext setting here as well (or at least inside of `DeserializeUsingBinaryFormatter`). That way BinaryFormatter gets trimmed by default, and users don't see ILLink warnings from this code.

Reply:
The idea being, if the user didn't enable `AllowBinaryFormatterUsageInTypeConverterSerialization`, you get an exception if you try to deserialize a binary formatted payload.